### ISSUE TITLE
Laser and stereo data misalignment (unrelated to diff between visual and collision geom)

### ISSUE THREAD

**Issue by osrf-migration** [osrf/drcsim#185]
**Original report ([archived issue](https://osrf-migration.github.io/drcsim-gh-pages/#!/osrf/drcsim/issues/185)) by Stefan Kohlbrecher (Bitbucket: [Stefan_Kohlbrecher](https://bitbucket.org/%7Bdb06b03e-7b31-4483-9018-534df5bd723f%7D/)).**

The original report had attachments: [atlas_lidar_misalignment.jpg](https://osrf-migration.github.io/drcsim-gh-pages/data/repositories/osrf/drcsim/issues/185/attachments/atlas_lidar_misalignment.jpg)

----------------------------------------
I have the latest drcsim-nightly installed and get sensor data as shown in the attached image. LIDAR points are consistently too short range when compared to stereo point clouds. As can be seen from the image and comparing the object's position in relation to the robot's feet, the stereo data seems to be correct, while the LIDAR data apparently is wrong.

To reproduce:

* Spawn a Atlas:
```
roslaunch atlas_utils atlas_sandia_hands.launch
```
* Let it look downwards (I did that using the joint dialog in gazebo)
* Make the LIDAR spin
```
rostopic pub --once /multisense_sl/set_spindle_speed std_msgs/Float64 '{data: 3.0}'
```
* Insert some objects
* Visualize both the stereo point cloud and LIDAR data with a decay time != 0 (for better visibility)

Any idea  on what is happening or even better, a fix, would be greatly appreciated.

**Comment by osrf-migration** [comment]
**Original comment by Thomas Koletschka (Bitbucket: [thomasko](https://bitbucket.org/%7Bb33092ac-6376-48d1-9803-d9e6de445d1a%7D/)).**

----------------------------------------
I can confirm this. LIDAR points seem to have a mean range that is 6-10cm closer (varies) to the sensor than it actually should be. I'm attaching two screenshots, one with the robot looking down at the ground with stereo correctly showing the ground plane at a mean height of ~0cm while lidar (red points) measures a height of ~10cm, and the other showing the measurement offset using a dumpster placed in front of the robot:

Attaching failed so here they are:

* http://i.imgur.com/0AWKs5l.png
* http://i.imgur.com/s5DG46B.png

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
Does anyone know when this appeared? (or more importantly if its going to take ~month to fix: what version of drcsim/gazebo didi it appear in?

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
This issue seems to have appeared in the last month. The version of gazebo I used in March didn't have this.

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
Might be related to [gazebo issue #503](https://bitbucket.org/osrf/gazebo/issue/503/), which suggests that laser ranges are measured from minimum range rather than the sensor origin.  That would produce scans that are too short.  In the case of the Hokuyo laser in the MultiSense-SL, each beam would be too short by 10cm.  If you substract 10cm from each beam, does the data look right?

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
I changed gazebo/sensors/RaySensor.cc to add on "+ 0.1;"
    double range = this->laserShape->GetRange(j * this->GetRayCount() + i) + 0.1;

... and that seems to be it. ground returns are correctly on the ground with this change and i also compared with some objects in the scene e.g. the vehicle.

Where is that set?

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
Here's a patch to try:

    diff -r 1a69666680ab gazebo/physics/MultiRayShape.cc
    --- a/gazebo/physics/MultiRayShape.cc	Sat Apr 13 18:17:01 2013 -0700
    +++ b/gazebo/physics/MultiRayShape.cc	Sun Apr 14 16:53:55 2013 -0700
    @@ -113,7 +113,7 @@
         gzthrow(stream.str());
       }
 
    -  return this->rays[_index]->GetLength();
    +  return this->GetRangeMin() + this->rays[_index]->GetLength(); 
     }
 
     //////////////////////////////////////////////////

It'll take a bit of work to get the tests updated to adapt to this change.

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
Could you describe how to patch this issue as its a pretty big regression for what's presumably a very straightforward fix.

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
thanks ;)

testing ....

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
My compilers say I don't have a GetRangeMin() method.

I dont work off of head. i'll need to download head and compile it to.

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------
Sorry, pasted the wrong version; take 2:

    diff -r 1a69666680ab gazebo/physics/MultiRayShape.cc
    --- a/gazebo/physics/MultiRayShape.cc	Sat Apr 13 18:17:01 2013 -0700
    +++ b/gazebo/physics/MultiRayShape.cc	Sun Apr 14 17:06:29 2013 -0700
    @@ -113,7 +113,7 @@
         gzthrow(stream.str());
       }
 
    -  return this->rays[_index]->GetLength();
    +  return this->rangeElem->GetValueDouble("min") + this->rays[_index]->GetLength(); 
     }
 
     //////////////////////////////////////////////////

That's still not quite right, as it produces ranges greater than the max.  Once I understand how the ray sensor works, I'll put in a proper fix.

**Comment by osrf-migration** [comment]
**Original comment by Maurice Fallon (Bitbucket: [mauricefallon](https://bitbucket.org/%7B8bca6a99-adab-444a-b7f0-81660c79adfb%7D/)).**

----------------------------------------
That seems to do the trick. The returns seem to be the same as adding on 0.1m

FYI: there is an method called GetMinRange() 
There was none called GetRangeMin()

Thanks! (signing off, its pretty late where i am)

**Comment by osrf-migration** [comment]
**Original comment by Brian Gerkey (Bitbucket: [Brian Gerkey](https://bitbucket.org/%7B3d8b5763-9dad-4be7-85c2-a7ab6cd49173%7D/), GitHub: [gerkey](https://github.com/gerkey)).**

----------------------------------------

* changed state from "new" to "resolved"

Fixed by [gazebo pull request #449](https://bitbucket.org/osrf/gazebo/pull-request/449), related to [gazebo issue #503](https://bitbucket.org/osrf/gazebo/issue/503).

Fix expected to appear in Gazebo 1.6.3 and 1.7
